Fix crash when user mapping has no username

erimatnor opened the pull request:
If a user mapping is created without a user name in the options, a
crash would happen when querying a distributed hypertable.

This change fixes the crash by assuming that the user to connect with
is the "current user" (i.e., no username mapping is done). This makes
it possible to create user mappings that only provide a password
option.

k-rus commented on tsl/src/remote/connection.c:
```diff
@@ -455,6 +455,11 @@ extract_connection_options(List *defelems, const char **keywords, const char **v
 	}
 
+	/* If a user was not found in the connection options, assume the current
+	 * user */
```
Since it is a proper sentence and starting with capitalized letter, it is good to end with a dot.
```suggestion
	 * user. */
```

k-rus commented on tsl/test/sql/dist_grant.sql:
```diff
@@ -262,7 +262,9 @@ SELECT * FROM create_distributed_hypertable('disttable_role_3', 'time', data_nod
 
 RESET ROLE;
-CREATE USER MAPPING FOR :ROLE_3 SERVER data2 OPTIONS (user :'ROLE_3', password :'ROLE_3_PASS');
+-- Create user mapping for ROLE_3, but don't specify user in
+-- options. The "current user" is assumed instead.
```
And what is the current user? Isn't it the super user used to connect to this database?

k-rus replied:
Actually not passing user means that `:ROLE_3` has no rights and thus the test fails no.

erimatnor replied:
I did see that test does not pass on Windows, so will investigate. But the test passes on other platforms and I don't understand what you mean by "no rights and thus the test fails"? Can you elaborate?

FWIW, the current user is what is returned by `GetUserId()` and (also reflected by) `SELECT current_user;`. AFAIK, user mappings associate no permissions with the user.

k-rus replied:
And the current user is not ROLE_3 to my understanding, thus the mapping doesn't work correctly, I guess. Strange that it behaves differently between Windows and Linux/MacOS.

erimatnor replied:
The current user is resolved when the connection is made, not when the mapping is created.

k-rus replied:
Is it possible to show the user mapping, e.g., which remote user will be used in the connection, in the test before setting the user to `:ROLE_3` and after?

erimatnor replied:
It is possible to show the contents of `pg_user_mapping`, but that only shows what was added when creating the mapping, which now doesn't include a user. Maybe we could print the connection options somehow, but we don't have a way to do it right now.

erimatnor replied:
I think I figured it out. Stupid mistake. I actually never added the user to the connection options.

k-rus reviewed (approved):
LGTM, but I don't follow the test change. See my comment.

k-rus reviewed (changes requested):
The change to test is incorrect.

pmwkaa commented on tsl/src/remote/connection.c:
```diff
@@ -455,6 +455,11 @@ extract_connection_options(List *defelems, const char **keywords, const char **v
+	if (NULL == *user)
+		*user = GetUserNameFromId(GetUserId(), true);
+
 	Assert(*user != NULL);
```
How this assert worked before?

erimatnor replied:
Not sure what you mean, but this is where the crash happened without my changes when assert was enabled. Without asserts (in release builds) it crashed later when it tried to md5-hash the user name to create a directory for a user SSL certs.

pmwkaa replied:
Oh okey, so it worked and crashed at that point. Makes sense then

mkindahl commented on tsl/src/remote/connection.c:
```diff
@@ -1129,6 +1140,23 @@ set_ssl_options(const char *user_name, const char **keywords, const char **value
 	*option_start = option_pos;
 }
 
+/*
+ * Finish the connection and, optionally, save the connection error.
+ */
+static void
+finish_connection(PGconn *conn, char **errmsg)
+{
+	if (NULL != errmsg)
+	{
+		if (NULL == conn)
+			*errmsg = "invalid connection";
+		else
+			*errmsg = pchomp(PQerrorMessage(conn));
+	}
+
+	PQfinish(conn);
+}
```
Wouldn't the code be easier to follow if this just raised an error if there were a connection problem (using `ereport`). Then it is not necessary to catch the error message and print it out, while anybody interested in catching the error can do so using `PG_TRY`.

With this approach we have different ways of handling errors: I think it would be better to stick to a single way.

erimatnor replied:
There's code that rely on this to not throw errors. Note also that throwing errors with libpq objects involved isn't always a good idea since those objects are malloc:ed and not managed through PostgreSQL's memory contexts.

mkindahl replied:
I don't understand.

You call `finish_connection` from inside `remote_connection_open_with_options_nothrow` and in each location where you call `remote_connection_open_with_options_nothrow` you immediately after check if it returns a null pointer and then you call `ereport`. If an `ereport` call inside `finish_connection` can break something, then it looks like the call immediately after `remote_connection_open_with_options_nothrow` would also break something. The only difference I see is that you pass the same memory upwards (in the `errmsg` variable), and then use it higher up in the call stack.

What am I missing?

erimatnor commented:
FYI, I added two more commits to this PR. First a small fix to polling the socket during connection establishment. In the previous PR, I had missed that one should check for write on the socket in the first iteration of the connection poll loop, so I changed that to fix flakiness on some platforms. Then I added a commit for adding detail errors during connection failures. This helped me debug but is also useful for the user since our errors are pretty opaque otherwise.

erimatnor commented on tsl/src/remote/connection.c:
```diff
@@ -1227,7 +1257,8 @@ remote_connection_open_with_options_nothrow(const char *node_name, List *connect
 					break;
-				default:
+				case PGRES_POLLING_ACTIVE:
+					Assert(false);
```
This event is not used but kept for backwards compatibility according to libpq comment.

erimatnor commented on tsl/src/remote/connection.c:
```diff
@@ -1189,13 +1215,17 @@ remote_connection_open_with_options_nothrow(const char *node_name, List *connect
 		if (PQstatus(pg_conn) == CONNECTION_BAD)
 		{
-			PQfinish(pg_conn);
+			finish_connection(pg_conn, errmsg);
```
This happens if malloc fails for pg_conn. Not really easy to test

k-rus commented on tsl/src/remote/connection.c:
```diff
@@ -1227,7 +1231,8 @@ remote_connection_open_with_options_nothrow(const char *node_name, List *connect
 					/* Should wait for write on socket */
 					events |= WL_SOCKET_WRITEABLE;
 					break;
-				default:
```
An example of PGconnectPoll usage in PG keeps `default` case and also have check on `pollstatus != PGRES_POLLING_OK`. Will it be good to keep the default case and handle it in the same way as `PGRES_POLLING_FAILED`?

k-rus commented on tsl/src/data_node.c:
```diff
@@ -552,16 +552,26 @@ connect_for_bootstrapping(const char *node_name, const char *const host, int32 p
 	}
-	return conn;
+
+	if (NULL == conn)
```
Is it possible that `conn != NULL` here? I suggest to remove this condition and throw the error.

k-rus commented on tsl/src/remote/connection.c:
```diff
@@ -1150,6 +1168,9 @@ remote_connection_open_with_options_nothrow(const char *node_name, List *connect
 	int option_count;
 	int option_pos;
 
+	if (NULL != errmsg)
+		*errmsg = NULL;
```
This seems to be unnecessary and instead should be initialised correctly by the caller.

k-rus commented on tsl/src/data_node.c:
```diff
@@ -552,16 +552,26 @@ connect_for_bootstrapping(const char *node_name, const char *const host, int32 p
+
+	if (NULL == conn)
+		ereport(ERROR,
+				(errcode(ERRCODE_SQLCLIENT_UNABLE_TO_ESTABLISH_SQLCONNECTION),
+				 errmsg("could not connect to \"%s\"", node_name),
+				 err == NULL ? 0 : errdetail("%s", err)));
```
Copy-pasted code with `connection.c`. I don't think it is important to improve now and I don't see an easy and clear pattern to fix.

k-rus reviewed (approved):
Should we actually disallow user mappings without user name specified? It is not dictated by PG and completely up to us and our FDW.

Otherwise looks OK. I have few comments, which will be good to address, except indicated one.
